man/flvmeta.1
.TH flvmeta 1 "April 2012" "flvmeta user manual"
.SH NAME
.PP
flvmeta - manipulate or extract metadata in Adobe Flash Video files
.SH SYNOPSIS
.PP
\f[B]flvmeta\f[] \f[I]INPUT_FILE\f[]
.PD 0
.P
.PD
\f[B]flvmeta\f[] \f[I]INPUT_FILE\f[] \f[I]OUTPUT_FILE\f[]
.PD 0
.P
.PD
\f[B]flvmeta\f[] \f[C]-D\f[]|\f[C]--dump\f[] [\f[I]options\f[]]
\f[I]INPUT_FILE\f[]
.PD 0
.P
.PD
\f[B]flvmeta\f[] \f[C]-F\f[]|\f[C]--full-dump\f[] [\f[I]options\f[]]
\f[I]INPUT_FILE\f[]
.PD 0
.P
.PD
\f[B]flvmeta\f[] \f[C]-C\f[]|\f[C]--check\f[] [\f[I]options\f[]]
\f[I]INPUT_FILE\f[]
.PD 0
.P
.PD
\f[B]flvmeta\f[] \f[C]-U\f[]|\f[C]--update\f[] [\f[I]options\f[]]
\f[I]INPUT_FILE\f[] [\f[I]OUTPUT_FILE\f[]]
.SH DESCRIPTION
.PP
\f[B]flvmeta\f[] is a command-line utility aimed at manipulating
Adobe(tm) Flash Video files (FLV), through several commands, only one of
which can be used for each invocation of the program.
.PP
It possesses the ability to compute and inject a variety of values in
the \f[I]onMetaData\f[] event tag, including keyframe indices used by
most video players to allow random-access seeking, notably for HTTP
pseudo-streamed files via a server-side module, by having the client
send the file offset looked up for the nearest desired keyframe.
.PD 0
.P
.PD
Tools such as \f[B]flvmeta\f[] must be used in the case the initial
encoding process is unable to inject those metadata.
.PP
It can also optionnally inject the \f[I]onLastSecond\f[] event, used to
signal the end of playback, for example to revert the player software to
a `stopped' state.
.PP
\f[B]flvmeta\f[] also has the ability to dump metadata and full file
information to standard output, in a variety of textual output formats,
including XML, YAML, and JSON.
.PP
Finally, the program can analyze FLV files to detect potential problems
and errors, and generate a textual report in a raw format, or in XML.
It has the ability to detect more than a hundred problems, going from
harmless to potentially unplayable, using a few real world encountered
issues.
This analysis can also determine and display the minimal Flash Player
version which can be used to correctly play a given file, as well as
codec information.
.PP
\f[B]flvmeta\f[] can operate on arbitrarily large files, and can handle
FLV files using extended (32-bit) timestamps.
It can guess video frame dimensions for all known video codecs supported
by the official FLV specification.
.PP
Its memory usage remains minimal, as it uses a two-pass reading
algorithm which permits the computation of all necessary tags without
loading anything more than the file's tags headers in memory.
Only the dumping of JSON data can consume more memory, since the bundled
\f[C]mjson\f[] library constructs a data tree in memory before
outputting the formatted data.
.SH COMMANDS
.PP
Only one command can be specified for an invocation of \f[B]flvmeta\f[].
The chosen command determines the mode of execution of the program.
.PP
By default, if no command is specified, \f[B]flvmeta\f[] will
implicitely choose the command to use according to the presence of
\f[I]INPUT_FILE\f[] and \f[I]OUTPUT_FILE\f[].
.PP
If only \f[I]INPUT_FILE\f[] is present, the \f[C]--dump\f[] command will
be executed.
.PP
If both \f[I]INPUT_FILE\f[] and \f[I]OUTPUT_FILE\f[] are present, the
\f[C]--update\f[] command will be executed.
.PP
Here is a list of the supported commands:
.SS -D, --dump
.PP
Dump a textual representation of the first \f[I]onMetaData\f[] tag found
in \f[I]INPUT_FILE\f[] to standard output.
The default format is XML, unless specified otherwise.
.PD 0
.P
.PD
It is also possible to specify another event via the \f[C]--event\f[]
option, such as \f[I]onLastSecond\f[].
.SS -F, --full-dump
.PP
Dump a textual representation of the whole contents of
\f[I]INPUT_FILE\f[] to standard output.
The default format is XML, unless specified otherwise.
.SS -C, --check
.PP
Print a report to standard output listing warnings and errors detected
in \f[I]INPUT_FILE\f[], as well as potential incompatibilities, and
informations about the codecs used in the file.
The exit code will be set to a non-zero value if there is at least one
error in the file.
.PP
The output format can either be plain text or XML using the
\f[C]--xml\f[] option.
It can also be disabled altogether using the \f[C]--quiet\f[] option if
you are only interested in the exit status.
.PP
Messages are divided into four specific levels of increasing importance:
.IP \[bu] 2
\f[B]info\f[]: informational messages that do not pertain to the file
validity
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]warning\f[]: messages that inform of oddities to the flv format but
that might not hamper file reading or playability, this is the default
level
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]error\f[]: messages that inform of errors that might render the
file impossible to play or stream correctly
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]fatal\f[]: messages that inform of errors that make further file
reading impossible therefore ending parsing completely
.PP
The \f[C]--level\f[] option allows to limit the display of messages to a
minimum level among those, for example if the user is only interested in
error messages and above.
.PP
Each message or message template presented to the user is identified by
a specific code of the following format:
.PP
\f[C][level][topic][id]\f[]
.IP \[bu] 2
\f[B]level\f[] is an upper-case letter that can be either I, W, E, F
according to the aforementioned message levels
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]topic\f[] is a two-digit integer representing the general topic of
the message
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]id\f[] is a unique three-digit identifier for the message, or
message template if parameterized
.PP
Messages can be related to the following topics :
.IP \[bu] 2
\f[B]10\f[] general flv file format
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]11\f[] file header
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]12\f[] previous tag size
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]20\f[] tag format
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]30\f[] tag types
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]40\f[] timestamps
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]50\f[] audio data
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]51\f[] audio codecs
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]60\f[] video data
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]61\f[] video codecs
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]70\f[] metadata
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]80\f[] AMF data
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]81\f[] keyframes
.PD 0
.P
.PD
.IP \[bu] 2
\f[B]82\f[] cue points
.PP
For example, represents a Warning in topic 51 with the id 050, which
represents a warning message related to audio codecs, in that case to
signal that an audio tag has an unknown codec.
.SS -U, --update
.PP
Update the given input file by inserting a computed \f[I]onMetaData\f[]
tag.
If \f[I]OUTPUT_FILE\f[] is specified, it will be created or overwritten
instead and the input file will not be modified.
If the original file is to be updated, a temporary file will be created
in the default temp directory of the platform, and it will be copied
over the original file at the end of the operation.
This is due to the fact that the output file is written while the
original file is being read due to the two-pass method.
.PP
The computed metadata contains among other data full keyframe
informations, in order to allow HTTP pseudo-streaming and random-access
seeking in the file.
.PP
By default, an \f[I]onLastSecond\f[] tag will be inserted, unless the
\f[C]--no-last-second\f[] option is specified.
.PP
Normally overwritten by the update process, the existing metadata found
in the input file can be preserved by the \f[C]--preserve\f[] option.
.PP
It is also possible to insert custom string values with the
\f[C]--add\f[] option, which can be specified multiple times.
.PP
By default, the update operation is performed without output, unless the
\f[C]--verbose\f[] option is specified, or the \f[C]--print-metadata\f[]
is used to print the newly written metadata to the standard output.
.SH OPTIONS
.SS DUMP
.TP
.B -d \f[I]FORMAT\f[], --dump-format=\f[I]FORMAT\f[]
specify dump format where \f[I]FORMAT\f[] is `xml' (default), `json',
`raw', or `yaml'.
Also applicable for the \f[B]--full-dump\f[] command.
.RS
.RE
.TP
.B -j, --json
equivalent to \f[B]--dump-format=json\f[]
.RS
.RE
.TP
.B -r, --raw
equivalent to \f[B]--dump-format=raw\f[]
.RS
.RE
.TP
.B -x, --xml
equivalent to \f[B]--dump-format=xml\f[]
.RS
.RE
.TP
.B -y, --yaml
equivalent to \f[B]--dump-format=yaml\f[]
.RS
.RE
.TP
.B -e \f[I]EVENT\f[], --event=\f[I]EVENT\f[]
specify the event to dump instead of \f[I]onMetaData\f[], for example
\f[I]onLastSecond\f[].
.RS
.RE
.SS CHECK
.TP
.B -l \f[I]LEVEL\f[], --level=\f[I]LEVEL\f[]
print only messages where level is at least \f[I]LEVEL\f[].
The levels are, by ascending importance, `info', `warning' (default),
`error', or `fatal'.
.RS
.RE
.TP
.B -q, --quiet
do not print messages, only return the status code
.RS
.RE
.TP
.B -x, --xml
generate an XML report instead of the default `compiler-friendly' text
.RS
.RE
.SS UPDATE
.TP
.B -m, --print-metadata
print metadata to stdout after update using the format specified by the
\f[B]--format\f[] option
.RS
.RE
.TP
.B -a \f[I]NAME=VALUE\f[], --add=\f[I]NAME=VALUE\f[]
add a metadata string value to the output file
.RS
.RE
.TP
.B -s, --no-lastsecond
do not create the \f[I]onLastSecond\f[] tag
.RS
.RE
.TP
.B -p, --preserve
preserve input file existing \f[I]onMetadata\f[] tags
.RS
.RE
.TP
.B -f, --fix
fix invalid tags from the input file
.RS
.RE
.TP
.B -i, --ignore
ignore invalid tags from the input file (the default behaviour is to
stop the update process with an error)
.RS
.RE
.TP
.B -t, --reset-timestamps
reset timestamps so \f[I]OUTPUT_FILE\f[] starts at zero.
This has been added because some FLV files are produced by cutting
bigger files, and the software doing the cutting does not resets the
timestamps as required by the standard, which can cause playback issues.
.RS
.RE
.TP
.B -k, \[em]all-keyframes
index all keyframe tags, including duplicate timestamps
.RS
.RE
.SS GENERAL
.TP
.B -v, --verbose
display informative messages
.RS
.RE
.TP
.B -V, --version
print version information and exit
.RS
.RE
.TP
.B -h, --help
display help on the program usage and exit
.RS
.RE
.SH FORMATS
.SH EXAMPLES
.SH EXIT STATUS
.SH BUGS
.PP
\f[B]flvmeta\f[] does not support encrypted FLV files yet.
.SH AUTHOR
.SH COPYRIGHT
.SH CONTACT


NEWS
Version 1.2.0 (2012-04-30)
  - Added proper command line handling and help.
  - Added the possibility to overwrite the input file
    when the output file is not specified or when both files
    are physically the same.
  - Added support for CMake builds in addition to autotools.
    It is now the official way to build flvmeta on Windows.
  - Added metadata and full file dumping, integrating former
    flvdump functionality into flvmeta.
  - Added support for XML, YAML, and JSON formats for dumping.
  - Added XML schemas describing the various formats used by flvmeta.
  - Added a file checking feature.
  - Added the possibility to print output file metadata after
    a successful update using one of the supported formats.
  - Added a feature to insert custom metadata strings while updating.
  - Added an option to disable insertion of the onLastSecond event.
  - Added an option to preserve existing metadata tags if possible.
  - Added an option to fix invalid tags while updating (this is
    a highly experimental feature, should be used with caution)
  - Added an option to ignore invalid tags as much as possible
    instead of exiting at the first error detected.
  - Added an option to reset the file timestamps in order to
    correctly start at zero, for example if the file has been
    incorrectly split by buggy tools.
  - Added an option to display informative messages while processing
    (not quite exhaustive for the moment).
  - Changed keyframe index generation so only the first keyframe
    for a given timestamp will be indexed. This behaviour can be
    overriden with the --all-keyframes/-k option.

Version 1.0.11 (2010-01-25)
  - Bug fix release.
  - Fixed video resolution detection when the first video frame
    is not a keyframe.
  - Fixed invalid timestamp handling in the case of decreasing timestamps.
  - Fixed AVC resolution computation when frame cropping rectangle
    is used.
  - Fixed handling of files with a non-zero starting timestamp.
  - Fixed datasize tag computation so only metadata are taken
    into account.

Version 1.0.10 (2009-09-02)
  - Bug fix release.
  - Fixed amf_data_dump numeric format.
  - Fixed extended timestamp handling.
  - Fixed video resolution detection causing a crash in the case
    the video tag data body has less data than required.

Version 1.0.9 (2009-06-23)
  - Bug fix release.
  - Fixed large file support so it will work for files bigger than 4GB.
  - Fixed date handling in AMF according to the official spec.
  - Fixed extended timestamp handling.
  - Fixed a bug causing reading invalid tags to potentially lead to
    memory overflow and creation of invalid files.

Version 1.0.8 (2009-05-08)
  - Added support for arbitrary large files (2GB+).
  - Added support for AVC (H.264) video resolution detection.
  - Added support for the Speex codec and rarely met video frame types.
  - Fixed a bug where two consecutive tags of different types and
    with decreasing timestamps would cause extended timestamps
    to be incorrectly used for the next tags.
  - Fixed a bug where zero would be used as video height and width
    if video resolution could not be detected.
  - Fixed a bug causing flvdump to crash after reading invalid tags.
    Flvdump now exits after the first invalid tag read.

Version 1.0.7 (2008-09-25)
  - Added support for extended timestamps.
    Now flvmeta can read and write FLV files longer than 4:39:37,
    as well as fix files with incorrect timestamps.
  - Added support for all codecs from the official specification.
  - Fixed a bug causing flvdump to lose track of tags in case
    of invalid metadata.

Version 1.0.6 (2008-05-28)
  - Bug fix release.
  - Fixed a flvdump crash under Linux caused by incorrect string
    handling.

Version 1.0.5 (2008-04-03)
  - Bug fix release.
  - Fixed a bug in the video size detection for VP60.
  - Added support for the AMF NULL type.
  - Simplified the AMF parser/writer.

Version 1.0.4 (2008-01-04)
  - Bug fix release.
  - Fixed a bug where some data tags wouldn't be written.
  - Fixed a date computation bug related to daylight saving.
  - Modified flvdump to make it more tolerant to malformed files.

Version 1.0.3 (2007-10-21)
  - Bug fix release.
  - Fixed major bugs in the AMF parser/writer.
  - Fixed a bug in the video size detection for VP6 alpha.
  - Fixed minor bugs.

Version 1.0.2 (2007-09-30)
  - Bug fix release.
  - Fixed issues on 64-bits architectures.
  - Fixed "times" metadata tag, which was incorrectly written
    as "timestamps".
  - Fixed audio delay computation.

Version 1.0.1 (2007-09-25)
  - Bug fix release.
  - Fixed a critical bug where file size and offsets would not
    be correctly computed if the input file did not have
    an onMetaData event.
  - Audio related metadata are not added anymore if the FLV file
    has no audio data.

Version 1.0 (2007-09-21)
  - This is the first public release.


src/flvmeta.c
/*
    $Id: flvmeta.c 231 2011-06-27 13:46:19Z marc.noirot $

    FLV Metadata updater

    Copyright (C) 2007-2012 Marc Noirot <marc.noirot AT gmail.com>

    This file is part of FLVMeta.

    FLVMeta is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    FLVMeta is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with FLVMeta; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "flvmeta.h"
#include "check.h"
#include "dump.h"
#include "update.h"

/*
    Command-line options
*/
static struct option long_options[] = {
    { "dump",               no_argument,        NULL, 'D'},
    { "full-dump",          no_argument,        NULL, 'F'},
    { "check",              no_argument,        NULL, 'C'},
    { "update",             no_argument,        NULL, 'U'},
    { "dump-format",        required_argument,  NULL, 'd'},
    { "json",               no_argument,        NULL, 'j'},
    { "raw",                no_argument,        NULL, 'r'},
    { "xml",                no_argument,        NULL, 'x'},
    { "yaml",               no_argument,        NULL, 'y'},
    { "event",              required_argument,  NULL, 'e'},
    { "level",              required_argument,  NULL, 'l'},
    { "quiet",              no_argument,        NULL, 'q'},
    { "print-metadata",     no_argument,        NULL, 'm'},
    { "add",                required_argument,  NULL, 'a'},
    { "no-lastsecond",      no_argument,        NULL, 's'},
    { "preserve",           no_argument,        NULL, 'p'},
    { "fix",                no_argument,        NULL, 'f'},
    { "ignore",             no_argument,        NULL, 'i'},
    { "reset-timestamps",   no_argument,        NULL, 't'},
    { "all-keyframes",      no_argument,        NULL, 'k'},
    { "verbose",            no_argument,        NULL, 'v'},
    { "version",            no_argument,        NULL, 'V'},
    { "help",               no_argument,        NULL, 'h'},
    { 0, 0, 0, 0 }
};

/* short options */
#define DUMP_COMMAND                "D"
#define FULL_DUMP_COMMAND           "F"
#define CHECK_COMMAND               "C"
#define UPDATE_COMMAND              "U"
#define DUMP_FORMAT_OPTION          "d:"
#define JSON_OPTION                 "j"
#define RAW_OPTION                  "r"
#define XML_OPTION                  "x"
#define YAML_OPTION                 "y"
#define EVENT_OPTION                "e:"
#define LEVEL_OPTION                "l:"
#define QUIET_OPTION                "q"
#define PRINT_METADATA_OPTION       "m"
#define ADD_OPTION                  "a:"
#define NO_LASTSECOND_OPTION        "s"
#define PRESERVE_OPTION             "p"
#define FIX_OPTION                  "f"
#define IGNORE_OPTION               "i"
#define RESET_TIMESTAMPS_OPTION     "t"
#define ALL_KEYFRAMES_OPTION        "k"
#define VERBOSE_OPTION              "v"
#define VERSION_OPTION              "V"
#define HELP_OPTION                 "h"

static void version(void) {
    printf("%s\n\n", PACKAGE_STRING);
    printf("%s\n", COPYRIGHT_STR);
    printf("This is free software; see the source for copying conditions. There is NO\n"
           "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n");
}

static void usage(const char * name) {
    fprintf(stderr, "Usage: %s [COMMAND] [OPTIONS] INPUT_FILE [OUTPUT_FILE]\n", name);
    fprintf(stderr, "Try `%s --help' for more information.\n", name);
}

static void help(const char * name) {
    printf("Usage: %s [COMMAND] [OPTIONS] INPUT_FILE [OUTPUT_FILE]\n", name);
    printf("\nIf OUTPUT_FILE is omitted for commands expecting it, INPUT_FILE will be overwritten instead.\n"
           "\nCommands:\n"
           "  -D, --dump                dump onMetaData tag (default without output file)\n"
           "  -F, --full-dump           dump all tags\n"
           "  -C, --check               check the validity of INPUT_FILE, returning 0 if\n"
           "                            the file is valid, or 10 if it contains errors\n"
           "  -U, --update              update computed onMetaData tag from INPUT_FILE\n"
           "                            into OUTPUT_FILE (default with output file)\n"
           /*    "  -A, --extract-audio       extract raw audio data into OUTPUT_FILE\n"*/
           /*    "  -E, --extract-video       extract raw video data into OUTPUT_FILE\n"*/
           "\nDump options:\n"
           "  -d, --dump-format=TYPE    dump format is of type TYPE\n"
           "                            TYPE is 'xml' (default), 'json', 'raw', or 'yaml'\n"
           "  -j, --json                equivalent to --dump-format=json\n"
           "  -r, --raw                 equivalent to --dump-format=raw\n"
           "  -x, --xml                 equivalent to --dump-format=xml\n"
           "  -y, --yaml                equivalent to --dump-format=yaml\n"
           "  -e, --event=EVENT         specify the event to be dumped instead of 'onMetadata'\n"
           "\nCheck options:\n"
           "  -l, --level=LEVEL         print only messages where level is at least LEVEL\n"
           "                            LEVEL is 'info', 'warning' (default), 'error', or 'fatal'\n"
           "  -q, --quiet               do not print messages, only return the status code\n"
           "  -x, --xml                 generate an XML report\n"
           "\nUpdate options:\n"
           "  -m, --print-metadata      print metadata to stdout after update using\n"
           "                            the specified format\n"
           "  -a, --add=NAME=VALUE      add a metadata string value to the output file\n"
           "  -s, --no-lastsecond       do not create the onLastSecond tag\n"
           "  -p, --preserve            preserve input file existing onMetadata tags\n"
           "  -f, --fix                 fix invalid tags from the input file\n"
           "  -i, --ignore              ignore invalid tags from the input file\n"
           "                            (the default is to stop with an error)\n"
           "  -t, --reset-timestamps    reset timestamps so OUTPUT_FILE starts at zero\n"
           "  -k, --all-keyframes       index all keyframe tags, including duplicate timestamps\n"
           "\nCommon options:\n"
           "  -v, --verbose             display informative messages\n"
           "\nMiscellaneous:\n"
           "  -V, --version             print version information and exit\n"
           "  -h, --help                display this information and exit\n");
    printf("\nPlease report bugs to <%s>\n", PACKAGE_BUGREPORT);
}

static int parse_command_line(int argc, char ** argv, flvmeta_opts * options) {
    int option, option_index;

    option_index = 0;
    do {
        option = getopt_long(argc, argv, 
            DUMP_COMMAND
            FULL_DUMP_COMMAND
            CHECK_COMMAND
            UPDATE_COMMAND
            DUMP_FORMAT_OPTION
            JSON_OPTION
            RAW_OPTION
            XML_OPTION
            YAML_OPTION
            EVENT_OPTION
            LEVEL_OPTION
            QUIET_OPTION
            PRINT_METADATA_OPTION
            ADD_OPTION
            NO_LASTSECOND_OPTION
            PRESERVE_OPTION
            FIX_OPTION
            IGNORE_OPTION
            RESET_TIMESTAMPS_OPTION
            ALL_KEYFRAMES_OPTION
            VERBOSE_OPTION
            VERSION_OPTION
            HELP_OPTION,
            long_options, &option_index);
        switch (option) {
            /*
                commands
            */
            case 'D':
                if (options->command != FLVMETA_DEFAULT_COMMAND) {
                    fprintf(stderr, "%s: only one command can be specified -- %s\n", argv[0], argv[optind]);
                    return EXIT_FAILURE;
                }
                options->command = FLVMETA_DUMP_COMMAND;
                break;
            case 'F':
                if (options->command != FLVMETA_DEFAULT_COMMAND) {
                    fprintf(stderr, "%s: only one command can be specified -- %s\n", argv[0], argv[optind]);
                    return EXIT_FAILURE;
                }
                options->command = FLVMETA_FULL_DUMP_COMMAND;
                break;
            case 'C':
                if (options->command != FLVMETA_DEFAULT_COMMAND) {
                    fprintf(stderr, "%s: only one command can be specified -- %s\n", argv[0], argv[optind]);
                    return EXIT_FAILURE;
                }
                options->command = FLVMETA_CHECK_COMMAND;
                break;
            case 'U':
                if (options->command != FLVMETA_DEFAULT_COMMAND) {
                    fprintf(stderr, "%s: only one command can be specified -- %s\n", argv[0], argv[optind]);
                    return EXIT_FAILURE;
                }
                options->command = FLVMETA_UPDATE_COMMAND;
                break;
            /*
                options
            */
            /* check options */
            case 'l':
                if (!strcmp(optarg, "info")) {
                    options->check_level = FLVMETA_CHECK_LEVEL_INFO;
                }
                else if (!strcmp(optarg, "warning")) {
                    options->check_level = FLVMETA_CHECK_LEVEL_WARNING;
                }
                else if (!strcmp(optarg, "error")) {
                    options->check_level = FLVMETA_CHECK_LEVEL_ERROR;
                }
                else if (!strcmp(optarg, "fatal")) {
                    options->check_level = FLVMETA_CHECK_LEVEL_FATAL;
                }
                else {
                    fprintf(stderr, "%s: invalid message level -- %s\n", argv[0], optarg);
                    usage(argv[0]);
                    return EXIT_FAILURE;
                }
                break;
            case 'q': options->quiet = 1; break;
            /* dump options */
            case 'd':
                if (!strcmp(optarg, "xml")) {
                    options->dump_format = FLVMETA_FORMAT_XML;
                }
                if (!strcmp(optarg, "raw")) {
                    options->dump_format = FLVMETA_FORMAT_RAW;
                }
                else if (!strcmp(optarg, "json")) {
                    options->dump_format = FLVMETA_FORMAT_JSON;
                }
                else if (!strcmp(optarg, "yaml")) {
                    options->dump_format = FLVMETA_FORMAT_YAML;
                }
                else {
                    fprintf(stderr, "%s: invalid output format -- %s\n", argv[0], optarg);
                    usage(argv[0]);
                    return EXIT_FAILURE;
                }
                break;
            case 'j': options->dump_format = FLVMETA_FORMAT_JSON;    break;
            case 'r': options->dump_format = FLVMETA_FORMAT_RAW;     break;
            case 'x':
                /* xml dump format, or generation of xml report */
                options->dump_format = FLVMETA_FORMAT_XML;
                options->check_xml_report = 1;
                break;
            case 'y': options->dump_format = FLVMETA_FORMAT_YAML;    break;
            case 'e': options->metadata_event = optarg;              break;
            /* update options */
            case 'm': options->dump_metadata = 1;                    break;
            case 'a':
                {
                    char * eq_pos;
                    eq_pos = strchr(optarg, '=');
                    if (eq_pos != NULL) {
                        *eq_pos = 0;
                        if (options->metadata == NULL) {
                            options->metadata = amf_associative_array_new();
                        }
                        amf_associative_array_add(options->metadata, optarg, amf_str(eq_pos + 1));
                    }
                    else {
                        fprintf(stderr, "%s: invalid metadata format -- %s\n", argv[0], optarg);
                        usage(argv[0]);
                        return EXIT_FAILURE;
                    }
                } break;
            case 's': options->insert_onlastsecond = 0;                  break;
            case 'p': options->preserve_metadata = 1;                    break;
            case 'f': options->error_handling = FLVMETA_FIX_ERRORS;      break;
            case 'i': options->error_handling = FLVMETA_IGNORE_ERRORS;   break;
            case 't': options->reset_timestamps = 1;                     break;
            case 'k': options->all_keyframes = 1;                        break;
            
            /*
                common options
            */
            case 'v': options->verbose = 1;  break;
            /*
                Miscellaneous
            */
            case 'V':
                options->command = FLVMETA_VERSION_COMMAND;
                return OK;
                break;
            case 'h':
                options->command = FLVMETA_HELP_COMMAND;
                return OK;
                break;
            /* last option */
            case EOF: break;
            /*
                error cases
            */
            /* unknown option */
            case '?':
            /* missing argument */
            case ':':
            /* we should not be here */
            default:
                usage(argv[0]);
                return EXIT_FAILURE;
        }
    } while (option != EOF);

    /* input filename */
    if (optind > 0 && optind < argc) {
        options->input_file = argv[optind];
    }
    else {
        fprintf(stderr, "%s: no input file\n", argv[0]);
        usage(argv[0]);
        return EXIT_FAILURE;
    }

    /* output filename */
    if (++optind < argc) {
        options->output_file = argv[optind];
    }

    /* determine command if default */
    if (options->command == FLVMETA_DEFAULT_COMMAND && options->output_file != NULL) {
        options->command = FLVMETA_UPDATE_COMMAND;
    }
    else if (options->command == FLVMETA_DEFAULT_COMMAND && options->output_file == NULL) {
        options->command = FLVMETA_DUMP_COMMAND;
    }

    /* output file is input file if not specified */
    if (options->output_file == NULL) {
        options->output_file = options->input_file;
    }

    return OK;
}

int main(int argc, char ** argv) {
    int errcode;

    /* flvmeta default options */
    static flvmeta_opts options;
    options.command = FLVMETA_DEFAULT_COMMAND;
    options.input_file = NULL;
    options.output_file = NULL;
    options.metadata = NULL;
    options.check_level = FLVMETA_CHECK_LEVEL_WARNING;
    options.quiet = 0;
    options.check_xml_report = 0;
    options.dump_metadata = 0;
    options.insert_onlastsecond = 1;
    options.reset_timestamps = 0;
    options.all_keyframes = 0;
    options.preserve_metadata = 0;
    options.error_handling = FLVMETA_EXIT_ON_ERROR;
    options.dump_format = FLVMETA_FORMAT_XML;
    options.verbose = 0;
    options.metadata_event = NULL;


    /* Command-line parsing */
    errcode = parse_command_line(argc, argv, &options);

    /* free metadata if necessary */
    if ((errcode != OK || options.command != FLVMETA_UPDATE_COMMAND) && options.metadata != NULL) {
        amf_data_free(options.metadata);
    }

    if (errcode == OK) {
        /* execute command */
        switch (options.command) {
            case FLVMETA_DUMP_COMMAND: errcode = dump_metadata(&options); break;
            case FLVMETA_FULL_DUMP_COMMAND: errcode = dump_flv_file(&options); break;
            case FLVMETA_CHECK_COMMAND: errcode = check_flv_file(&options); break;
            case FLVMETA_UPDATE_COMMAND: errcode = update_metadata(&options); break;
            case FLVMETA_VERSION_COMMAND: version(); break;
            case FLVMETA_HELP_COMMAND: help(argv[0]); break;
        }

        /* error report */
        switch (errcode) {
            case ERROR_OPEN_READ: fprintf(stderr, "%s: cannot open %s for reading\n", argv[0], options.input_file); break;
            case ERROR_NO_FLV: fprintf(stderr, "%s: %s is not a valid FLV file\n", argv[0], options.input_file); break;
            case ERROR_EOF: fprintf(stderr, "%s: unexpected end of file\n", argv[0]); break;
            case ERROR_MEMORY: fprintf(stderr, "%s: memory allocation error\n", argv[0]); break;
            case ERROR_EMPTY_TAG: fprintf(stderr, "%s: empty FLV tag\n", argv[0]); break;
            case ERROR_OPEN_WRITE: fprintf(stderr, "%s: cannot open %s for writing\n", argv[0], options.output_file); break;
            case ERROR_INVALID_TAG: fprintf(stderr, "%s: invalid FLV tag\n", argv[0]); break;
            case ERROR_WRITE: fprintf(stderr, "%s: unable to write to %s\n", argv[0], options.output_file); break;
        }
    }

    return errcode;
}
